### PR TITLE
[Solver PR 3] Update solvers from master

### PR DESCRIPTION
**Pull request by keveleigh**
Overview
---
The InBetween solver was missing, and a few were missing features. This also ports over the custom inspectors from master.

This PR does not yet re-add the ControllerFinder script and the ability to use controllers as the tracked object for a solver. That's coming in the next PR.

This fixes some issues where lines were moved into `OnValidate` calls, but the fields being set weren't serialized and thus were getting lost on play.

Related to #2465

Ports over #2222 and #2245

### CODE REVIEW

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs [hunk 1]
```diff
@@ -82,17 +142,63 @@ private void Update()
...
+                    }
                 }
             }
         }
 
+        protected void OnDestroy()
+        {
+            if (transformWithOffset != null)
+            {
+                Destroy(transformWithOffset);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: we should check if we're in the editor here and destroy immediate.

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs [hunk 2]
```diff
@@ -82,17 +142,63 @@ private void Update()
...
+                case TrackedObjectType.Head:
+                    TrackTransform(CameraCache.Main.transform);
+                    break;
+                // Other cases will come online as ControllerFinder is ported appropriately.
+            }
+        }
+
+        private void TrackTransform(Transform newTrackedTransform)
+        {
+            TransformTarget = RequiresOffset() ? MakeOffsetTransform(newTrackedTransform) : newTrackedTransform;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: should `RequiresOffset()` be a property?

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs [hunk 3]
```diff
@@ -82,17 +142,63 @@ private void Update()
...
+        }
+
+        private bool RequiresOffset()
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: Yeah we can make this a property.

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs [hunk 4]
```diff
@@ -155,13 +156,34 @@ private Vector3 RaycastDirection
...
+        private float ScaleOverride => useLinkedAltScaleOverride ? SolverHandler.AltScale.Current.magnitude : volumeCastSizeOverride;
 
-        private void OnValidate()
+        protected override void OnValidate()
         {
+            base.OnValidate();
+
             if (raycastMode == RaycastModeType.Box)
             {
-                boxCollider = gameObject.EnsureComponent<BoxCollider>();
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: Why remove this and add error for developer to manually add one?

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/InBetween.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        /// </summary>
+        public float PartwayOffset
+        {
+            get { return partwayOffset; }
+            set { partwayOffset = Mathf.Clamp(value, 0.0f, 1.0f); }
+        }
+
+        [SerializeField]
+        [Tooltip("Tracked object to calculate position and orientation for the second object. If you want to manually override and use a scene object, use the TransformTarget field.")]
+        [HideInInspector]
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/InBetween.cs]: Why have this attibute when we have a custom inspector?

File: Assets/MixedRealityToolkit-SDK/Inspectors/Utilities/Solvers/InBetweenInspector.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+namespace Microsoft.MixedReality.Toolkit.SDK.Inspectors.Utilities.Solvers
+{
+    [CustomEditor(typeof(InBetween))]
+    public class InBetweenEditor : Editor
+    {
+        private SerializedProperty trackedObjectProperty;
+        private SerializedProperty transformTargetProperty;
+        private InBetween solverInBetween;
+
+        private static readonly string[] fieldsToExclude = new string[] { "m_Script" };
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Inspectors/Utilities/Solvers/InBetweenInspector.cs]: Why do we need this if we're already getting the serialized properties?  I don't see us calling the `base.OnInspectorGUI`

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/InBetween.cs (hunk 5 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/InBetween.cs]: You didn't like the way we were hiding this field in the original custom inspector when this was added to master in #2245, so this tag was added. I'm happy to revert.

File: Assets/MixedRealityToolkit/_Core/Definitions/Utilities/TrackedObjectType.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    public enum TrackedObjectType
+    {
+        /// <summary>
+        /// Calculates position and orientation from the main camera.
+        /// </summary>
+        Head = 0,
+        /// <summary>
+        /// Calculates position and orientation from the left motion-tracked controller.
+        /// </summary>
+        MotionControllerLeft,
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Utilities/TrackedObjectType.cs]: Maybe we should be more platform agnostic here.

`TrackedControllerLeft`?

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs (hunk 1 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: What's the benefit? Unity recommends `Destroy` only for game code, which this is.

https://docs.unity3d.com/ScriptReference/Object.DestroyImmediate.html

>You are strongly recommended to use Destroy instead.

>In game code you should use Object.Destroy instead.

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs (hunk 4 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: A few reasons:

1. `AddComponent` (called by `EnsureComponent`) cannot be called in `OnValidate`.
2. Why would we potentially add a second, unwanted collider instead of informing the developer to add it correctly?

image

File: Assets/MixedRealityToolkit-SDK/Inspectors/Utilities/Solvers/InBetweenInspector.cs (hunk 6 above)
Comment (Ecnassianer) [reply Assets/MixedRealityToolkit-SDK/Inspectors/Utilities/Solvers/InBetweenInspector.cs]: Line 43:
DrawPropertiesExcluding(serializedObject, fieldsToExclude);

File: Assets/MixedRealityToolkit/_Core/Definitions/Utilities/TrackedObjectType.cs (hunk 7 above)
Comment (Ecnassianer) [reply Assets/MixedRealityToolkit/_Core/Definitions/Utilities/TrackedObjectType.cs]: Microsoft, Valve, and Oculus all use "motion controller" as a generic term to refer to their devices. Seems pretty platform agnostic to me, and clearer than Tracked Controller imo.

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs (hunk 1 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: In editor if you call Destroy, it'll throw warnings and suggest calling destroy immediate.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: Which is why checking if `Application.isEditor` before calling the appropriate method.

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs (hunk 4 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: good points.

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs (hunk 3 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: Updated

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs (hunk 1 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: This isn't editor code though. When will this be invoked in edit mode?

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs [hunk 8]
```diff
@@ -13,7 +15,53 @@ namespace Microsoft.MixedReality.Toolkit.SDK.Utilities.Solvers
     public class SolverHandler : MonoBehaviour
     {
         [SerializeField]
-        [Tooltip("The target transform that the solvers will act upon.")]
+        [Tooltip("Tracked object to calculate position and orientation from. If you want to manually override and use a scene object, use the TransformTarget field.")]
+        private TrackedObjectType trackedObjectToReference = TrackedObjectType.Head;
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: Missing summary (tooltip text should go here)

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/Solver.cs [hunk 9]
```diff
@@ -39,39 +46,41 @@ public abstract class Solver : MonoBehaviour
         /// <summary>
         /// The handler reference for this solver that's attached to this <see cref="GameObject"/>
         /// </summary>
+        [SerializeField]
+        [HideInInspector]
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/Solver.cs]: With a custom inspector, so we still need this?

Comment (keveleigh) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/Solver.cs]: Where's the custom inspector?

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs [hunk 10]
```diff
@@ -13,7 +15,53 @@ namespace Microsoft.MixedReality.Toolkit.SDK.Utilities.Solvers
...
+            {
+                additionalOffset = value;
+                TransformTarget = MakeOffsetTransform(TransformTarget);
+            }
+        }
+
+        [SerializeField]
+        [Tooltip("Add an additional rotation on top of the tracked object. Useful for tracking what is essentially the up or right/left vectors.")]
+        private Vector3 additionalRotation;
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: also missing tooltip summary

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs [hunk 11]
```diff
@@ -13,7 +15,53 @@ namespace Microsoft.MixedReality.Toolkit.SDK.Utilities.Solvers
...
+                    TransformTarget = null;
+                    AttachToNewTrackedObject();
+                }
+            }
+        }
+
+        [SerializeField]
+        [Tooltip("Add an additional offset of the tracked object to base the solver on. Useful for tracking something like a halo position above your head or off the side of a controller.")]
+        private Vector3 additionalOffset;
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: summary

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs [hunk 12]
```diff
@@ -82,17 +144,58 @@ private void Update()
...
+        protected void OnDestroy()
+        {
+            if (transformWithOffset != null)
+            {
+                Destroy(transformWithOffset);
+            }
+        }
+
         #endregion MonoBehaviour Implementation
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SolverHandler.cs]: summary

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs [hunk 13]
```diff
@@ -79,35 +78,37 @@ private enum OrientModeEnum
...
+        [Tooltip("Orientation mode. None = no orienting, Vertical = Face head, but always oriented up/down, Full = Aligned to surface normal completely")]
         private OrientModeEnum orientationMode = OrientModeEnum.Vertical;
 
         [SerializeField]
         [Tooltip("Orientation Blend Value 0.0 = All head 1.0 = All surface")]
-        private float orientBlend = 0.65f;
+        private float orientationBlend = 0.65f;
 
         [SerializeField]
         private bool debugEnabled = false;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: missing summary

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs [hunk 14]
```diff
@@ -517,10 +570,9 @@ private void FindPlacementPlane(Vector3 origin, Vector3 direction, Vector3[] pos
             }
 
             // Figure out how far the plane should be.
-            if (!useClosestDistance && closestPoint >= 0.0f)
+            if (!useClosestDistance && closestPoint >= 0)
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: why this change? Is it now an integer?

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs [hunk 15]
```diff
@@ -517,10 +570,9 @@ private void FindPlacementPlane(Vector3 origin, Vector3 direction, Vector3[] pos
...
             {
                 float centerPlaneDistance;
-
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: seems like a rather unnecessary formatting change

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs [hunk 16]
```diff
@@ -177,7 +199,6 @@ private Quaternion CalculateMagnetismOrientation(Vector3 direction, Vector3 surf
         {
             // Calculate the surface rotation
             Vector3 newDirection = -surfaceNormal;
-
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: seems like a rather unnecessary formatting change

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs [hunk 17]
```diff
@@ -79,35 +78,37 @@ private enum OrientModeEnum
...
         /// the head facing direction, from the eye to the object, or to the solver's linked position (working from
         /// the previous solvers)
         /// </summary>
         private Vector3 RaycastEndPoint
         {
             get
             {
+                Vector3 origin = RaycastOrigin;
                 Vector3 endPoint = Vector3.forward;
-
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: seems like a rather unnecessary formatting change

File: Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs (hunk 15 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: I don't consider mrtk_development to be the source truth for solvers, as they don't work in their current state.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: Right, and changing these lines isn't required to get it to work.

Comment (Ecnassianer) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: This makes the code more readable, and since we're in the middle of a port, it seems like the perfect time to polish up. I'm usually pretty against unnecessary code churn, but I don't think this should be reverted.

Comment (keveleigh) [reply Assets/MixedRealityToolkit-SDK/Features/Utilities/Solvers/SurfaceMagnetism.cs]: Good points from both of you. I unreverted these in my last commit.
